conductor-client/src/main/kotlin/com/openlattice/data/jobs/DataDeletionJob.kt

```diff
@@ -286,36 +327,50 @@ class DataDeletionJob(
         ps.addBatch()
     }
 
+    //TODO: entity key ids should be paired with their entity key ids :-/
     @JsonIgnore
     private fun deletePropertyOfEntityFromS3(
-            entitySetIds: Collection<UUID>,
-            entityKeyIds: Collection<UUID>,
-            propertyTypeIds: Collection<UUID>
+        entitySetIds: Collection<UUID>,
+        entityKeyIds: Collection<UUID>,
+        propertyTypeIds: Collection<UUID>
     ) {
-        val s3Keys = BasePostgresIterable<String>(
-                PreparedStatementHolderSupplier(hds, selectEntitiesTextProperties, FETCH_SIZE) { ps ->
-                    val connection = ps.connection
-                    val entitySetIdsArr = PostgresArrays.createUuidArray(connection, entitySetIds)
-                    val propertyTypeIdsArr = PostgresArrays.createUuidArray(connection, propertyTypeIds)
-                    val entityKeyIdsArr = PostgresArrays.createUuidArray(connection, entityKeyIds)
-                    ps.setArray(1, entitySetIdsArr)
-                    ps.setArray(2, propertyTypeIdsArr)
-                    ps.setArray(3, entityKeyIdsArr)
+
+        val s3Keys =
+            entitySetIds
+                .groupBy(resolver::getDataSourceName)
+                .flatMap { (dataSourceName, entitySetIdsForDataSource) ->
+                    val hds = resolver.getDataSource(dataSourceName)
+                    BasePostgresIterable<String>(
+                        PreparedStatementHolderSupplier(hds, selectEntitiesTextProperties, FETCH_SIZE) { ps ->
+                            val connection = ps.connection
+                            val entitySetIdsArr = PostgresArrays.createUuidArray(connection, entitySetIdsForDataSource)
+                            val propertyTypeIdsArr = PostgresArrays.createUuidArray(connection, propertyTypeIds)
+                            val entityKeyIdsArr = PostgresArrays.createUuidArray(connection, entityKeyIds)
+                            ps.setArray(1, entitySetIdsArr)
+                            ps.setArray(2, propertyTypeIdsArr)
+                            ps.setArray(3, entityKeyIdsArr)
+                        }
+                    ) { it.getString(getMergedDataColumnName(PostgresDatatype.TEXT)) }
```

Comment:
No need for the `.toList()`?

Reply:
Previously that iterable wouldn't get evaluated until the toList() was called. In this case, I'm assuming the flatMap triggers the evaluation of all the iterables and concatenating tem.